imap_processing/mag/l0/decom_mag.py

```diff
@@ -71,58 +71,118 @@ def export_to_xarray(l0_data: list[MagL0]):
     ----------
     l0_data: list[MagL0]
         A list of MagL0 datapoints
+
+    Returns
+    -------
+    norm_data : xr.Dataset
+        xarray dataset for generating burst data CDFs
+    burst_data : xr.Dataset
+        xarray dataset for generating burst data CDFs
     """
     # TODO split by mago and magi using primary sensor
     # TODO split by norm and burst
-    norm_data = defaultdict(list)
-    burst_data = norm_data.copy()
+    norm_data = []
+    burst_data = []
 
     for datapoint in l0_data:
         if datapoint.ccsds_header.PKT_APID == Mode.NORMAL:
-            for key, value in dataclasses.asdict(datapoint).items():
-                if key != "ccsds_header":
-                    norm_data[key].append(value)
+            norm_data.append(datapoint)
         if datapoint.ccsds_header.PKT_APID == Mode.BURST:
-            burst_data["SHCOARSE"].append(datapoint.SHCOARSE)
-            burst_data["raw_vectors"].append(datapoint.VECTORS)
+            burst_data.append(datapoint)
+
+    norm_dataset = None
+    burst_dataset = None
+
+    if len(norm_data) > 0:
+        norm_dataset = generate_dataset(norm_data)
+    if len(burst_data) > 0:
+        burst_dataset = generate_dataset(burst_data)
+
+    return norm_dataset, burst_dataset
+
+
+def generate_dataset(l0_data: list[MagL0]):
+    """
+    Generate a CDF dataset from the sorted L0 MAG data.
+
+    Used to create 2 similar datasets, for norm and burst data.
+
+    Parameters
+    ----------
+    l0_data : list[MagL0]
+        List of sorted L0 MAG data.
+
+    Returns
+    -------
+    dataset : xr.Dataset
+        xarray dataset with proper CDF attributes and shape.
+    """
+    vector_data = np.zeros((len(l0_data), len(l0_data[0].VECTORS)))
```

Comment:
Rather than padding below, you might be able to calculate it upfront here with a max.
```suggestion
    vector_data = np.zeros((len(l0_data), max(lambda x: len(x.VECTORS) for x in l0_data))
```

Reply:
That requires looping through all the data, whereas what I have loops through once and takes the performance hit on resizing the np array. I think the array is not likely to get expanded very often, so this is better performance overall (also, because np in general is way faster than normal python stuff)

Reply:
:+1: Maybe a question for another time then is why are we getting resizes at all, is this expected? I would have figured the packets to all contain X number of vectors.

Reply:
No, the vector count can vary based on the settings of the instrument

Reply:
Interesting, I would have expected it to only be a difference between burst and normal mode, but otherwise, be measured at the same frequency/cadence.

Another follow-on here then is whether `0` is an OK fill value for this data, or do you want `np.nan` to indicate that the entry didn't have any vector measurements past a certain point?

I guess I'm not entirely clear on what the shifting shape means. I think the shape is `(npackets, ndirections)` by looking later at the xarray dataset, but is it OK to have a shifting directions variable, or do we need different `ndirections` variables within the same dataset to distinguish these? A linearly spaced ndirections instead of only at the beginning (padding 0s to the end)?

I might be getting confused with thinking about higher level data products though, and just don't understand enough about the lower level organization to expect here. So just keep these questions in mind, but you don't need to address anything now.

Reply:
I agree that "direction" is not the correct axis for this particular CDF. It should be something else, with "Direction" being every other data product. But, I am planning to discuss this with mag when I get back. They didn't reply with their thoughts before now, so it will have to stay as my first draft for now 🤷 I asked if 0 was an ok fill value and they said yes. The byte vectors are varying lengths but you can calculate the number of fixed-length vectors from other data so the length per packet is known when processing.